# AfterWork Data Science: Data Wrangling with Python Project

## 1. Defining the Question

### a) Specifying the Data Analysis Question

Recommendations on reducing cases of breakdown of buses in the city of New York

### b) Defining the Metric for Success

We will have accomplished our objective if we provide suggestions that decrement incidences in bus breakdown 

### c) Understanding the context 

The Bus Breakdown and Delay system
collects information from school bus vendors operating in the field in real-time. Bus staff that
encounter delays during the route are instructed to radio the dispatcher at the bus vendor’s
central office. The bus vendor staff log into the Bus Breakdown and Delay system to record the
event and notify OPT. OPT customer service agents use this system to inform parents who call
with questions regarding bus service. The Bus Breakdown and Delay system is publicly
accessible and contains real-time updates. All information in the system is by school bus vendor
staff.


### d) Recording the Experimental Design

Load libraries and datasets.

Select the most appropriate variables to work with.

Find and deal with duplicated data.

Find and deal with missing values.

Drop data that does not match 

Perform Data analysis to answer questions


### e) Data Relevance

The given data sets were relevant in answering the research question.

## 2. Reading the Data

In [ ]:
# Importing our libraries
# ---
import pandas as pd # library for data manipulation
import numpy as np  # library for scientific computations.


In [ ]:
# Load the data below
# --- 
# Dataset url = : https://bit.ly/BusBreakdownDataset

BusBreakdown_df=pd.read_csv("https://bit.ly/BusBreakdownDataset")
BusBreakdown_df

School_Year  ...  School_Age_or_PreK
0        2015-2016  ...          School-Age
1        2015-2016  ...          School-Age
2        2015-2016  ...               Pre-K
3        2015-2016  ...          School-Age
4        2015-2016  ...          School-Age
...            ...  ...                 ...
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[281110 rows x 21 columns]

In [ ]:
# Checking the first 5 rows of data
# ---
BusBreakdown_df.head()

School_Year  Busbreakdown_ID  ... Breakdown_or_Running_Late School_Age_or_PreK
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [ ]:
# Checking the last 5 rows of data
# ---
#
BusBreakdown_df.tail()

School_Year  ...  School_Age_or_PreK
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[5 rows x 21 columns]

In [ ]:
# Sample 10 rows of data
# ---
BusBreakdown_df.sample(10)

School_Year  ...  School_Age_or_PreK
214601   2017-2018  ...          School-Age
70726    2016-2017  ...          School-Age
179385   2015-2016  ...          School-Age
104057   2016-2017  ...          School-Age
150775   2017-2018  ...          School-Age
113396   2017-2018  ...          School-Age
13125    2015-2016  ...          School-Age
109630   2017-2018  ...          School-Age
100995   2016-2017  ...          School-Age
141990   2018-2019  ...          School-Age

[10 rows x 21 columns]

In [ ]:
# Checking number of rows and columns
# ---
BusBreakdown_df.shape

(281110, 21)

In [ ]:
# Checking datatypes
# ---
BusBreakdown_df.dtypes

Record your observations below:

*   Observation 1 - Collected data has 281,110 rows and 21 columns

*   Observation 2- Data has missing values



## 3. External Data Source Validation

https://council.nyc.gov/data/school-bus-delays/

## 4. Data Preparation

### Performing Data Cleaning

In [ ]:
# Checking datatypes and missing entries of all the variables
# ---

#find any missing  entries
print(BusBreakdown_df.isnull().any())

#Dropping records with missing values
BusBreakdown_notnull_df = BusBreakdown_df.dropna()

BusBreakdown_notnull_df.isnull().sum()




School_Year                        False
Busbreakdown_ID                    False
Run_Type                            True
Bus_No                              True
Route_Number                        True
Reason                              True
Schools_Serviced                    True
Occurred_On                        False
Created_On                         False
Boro                                True
Bus_Company_Name                   False
How_Long_Delayed                    True
Number_Of_Students_On_The_Bus      False
Has_Contractor_Notified_Schools    False
Has_Contractor_Notified_Parents    False
Have_You_Alerted_OPT               False
Informed_On                        False
Incident_Number                     True
Last_Updated_On                    False
Breakdown_or_Running_Late          False
School_Age_or_PreK                 False
dtype: bool


School_Year                        0
Busbreakdown_ID                    0
Run_Type                           0
Bus_No                             0
Route_Number                       0
Reason                             0
Schools_Serviced                   0
Occurred_On                        0
Created_On                         0
Boro                               0
Bus_Company_Name                   0
How_Long_Delayed                   0
Number_Of_Students_On_The_Bus      0
Has_Contractor_Notified_Schools    0
Has_Contractor_Notified_Parents    0
Have_You_Alerted_OPT               0
Informed_On                        0
Incident_Number                    0
Last_Updated_On                    0
Breakdown_or_Running_Late          0
School_Age_or_PreK                 0
dtype: int64

We observe the following from our dataset:

*   Observation 1- There are no duplicates in the dataset

*   Observation 2- there are missing values in multiple columns



In [ ]:
# Standardizing your dataset 
# ---
# 
BusBreakdown_df.columns
BusBreakdown_df_Columns = BusBreakdown_df.columns.str.strip()

In [ ]:
# Checking how many duplicate rows are there in the data
# ---

BusBreakdown_df_dup = BusBreakdown_df[BusBreakdown_df.duplicated()]
BusBreakdown_df_dup

#check duplicates values in the Bus_breakdownID column
BusBreakdown_df_dupID= BusBreakdown_df[BusBreakdown_df.duplicated(['Busbreakdown_ID'])]
BusBreakdown_df_dupID

Empty DataFrame
Columns: [School_Year, Busbreakdown_ID, Run_Type, Bus_No, Route_Number, Reason, Schools_Serviced, Occurred_On, Created_On, Boro, Bus_Company_Name, How_Long_Delayed, Number_Of_Students_On_The_Bus, Has_Contractor_Notified_Schools, Has_Contractor_Notified_Parents, Have_You_Alerted_OPT, Informed_On, Incident_Number, Last_Updated_On, Breakdown_or_Running_Late, School_Age_or_PreK]
Index: []

We observe the following from our dataset:

*   Observation 1-There are no duplicate records in the dataset

*   Observation 2- There are no duplicates of the Bus_Breakdown_ID



In [ ]:
# Checking if any of the columns are all null
# ---

empty_cols = [col for col in BusBreakdown_df if BusBreakdown_df[col].isnull().all()]
empty_cols

[]

We observe the following from our dataset:

*   Observation 1-No column is all null




In [ ]:
# Checking if any of the rows are all null
# ---
#
print(BusBreakdown_df.isnull().values.all())



False


We observe the following from our dataset:

*   Observation -No row is all null




In [ ]:
# Checking if the "Yes/No" fields contain only these 2 values
# for have_you_alerted_opt variable
# ---
# Hint: Use unique() function
#

Alerted_OPT= BusBreakdown_df['Have_You_Alerted_OPT'].unique() 

list(Alerted_OPT)


['Yes', 'No']

We observe the following from our dataset:

* Have_You_Alerted_OPT column contains only two values, Yes and No



In [ ]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_parents variable
# ---
# 


Contractor_Notified_parents= BusBreakdown_df['Has_Contractor_Notified_Parents'].unique()

list(Contractor_Notified_parents)

['No', 'Yes']

We observe the following from our dataset:

* Has_Contractor_Notified_Parents  column contains only two values, Yes and No



In [ ]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_schools variable
# ---
#  

Contractor_Notified_school= BusBreakdown_df['Has_Contractor_Notified_Schools'].unique()

list(Contractor_Notified_school)

['Yes', 'No']

We observe the following from our dataset:

Has_Contractor_Notified_Schools column contains only two values, Yes and No



In [ ]:
# Checking unique values in break_down_or_running_late variable to ensure there is no duplication
# ---
#

Breakdown_Late=BusBreakdown_df['Breakdown_or_Running_Late'].unique()

list(Breakdown_Late)

['Running Late', 'Breakdown']

In [ ]:
# Checking unique values in school_age_or_prek variable
# ---
# 

School_Age_PreK=BusBreakdown_df['School_Age_or_PreK'].unique()

list(School_Age_PreK)



['School-Age', 'Pre-K']

In [ ]:
# Checking unique values in school_year variable
# ---
#  
BusBreakdown_df.columns
Years=BusBreakdown_df['School_Year'].unique()

list(Years)

['2015-2016', '2017-2018', '2018-2019', '2016-2017', '2019-2020']

In [ ]:
# Checking unique values in reason variable
# ---
#  

Reason = BusBreakdown_df['Reason'].unique()

list(Reason)



['Heavy Traffic',
 'Flat Tire',
 'Other',
 'Won`t Start',
 'Mechanical Problem',
 'Problem Run',
 'Accident',
 'Late return from Field Trip',
 'Delayed by School',
 'Weather Conditions',
 nan]

In [ ]:
# Checking unique values in run_type variable
# ---
#  

RunType= BusBreakdown_df['Run_Type'].unique()

list(RunType)

['Special Ed AM Run',
 'Pre-K/EI',
 'General Ed AM Run',
 'General Ed Field Trip',
 'Special Ed PM Run',
 'General Ed PM Run',
 'Project Read PM Run',
 'Special Ed Field Trip',
 'Project Read AM Run',
 'Project Read Field Trip',
 nan]

In [ ]:
# Checking unique values in boro variable
# ---
#   

BusBreakdown_df.columns

Boro= BusBreakdown_df['Boro'].unique()

list(Boro)

['New Jersey',
 'Manhattan',
 'Bronx',
 'Westchester',
 'Brooklyn',
 'Rockland County',
 'Nassau County',
 nan,
 'Queens',
 'Staten Island',
 'Connecticut',
 'All Boroughs']

### Overall Data Cleaning Observations
**Missing Values**

- There are a large number of missing values in the fields "How_Long_Delayed" which is important to our analysis. 
- There is an extremely large number of missing values in the "Incident_number" field but this is not incidental to our analysis and cannot be filled in without additional information.

**Error in values**

- "How_Long_Delayed" contains string values such as "MINS" or "mins" and a range of values, which needs to be changed to single integer value for our analysis.

**Error in Datatypes**

- "How_Long_Delayed" is a string datatype, should be converted to integer type.

**Error in field names**
- The column name "Boro" should be renamed to "Borough".


### Next Steps: Data Cleaning Steps

**Error in values**

- Extract the first integer value (lowest delay time) in the column "How_Long_Delayed"


**Missing Values**

- Impute the missing values in the field "How_Long_Delayed" with the mean value.


**Error in Datatypes**

- Convert "How_Long_Delayed" to int datatype.



**Error in field names**

- Rename the column "Boro" to "Borough".

In [ ]:
# Lets first start by creating a copy of our dataframe 
# df_clean = df.copy(). We will use this copy as our cleaning copy.
# ---
#
BusBreakdown_df_clean = BusBreakdown_df.copy()
BusBreakdown_df_clean.head()

School_Year  Busbreakdown_ID  ... Breakdown_or_Running_Late School_Age_or_PreK
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [ ]:
# Then extracting the lowest delay time in the column how_long_delayed from the string
# 
BusBreakdown_df_clean['How_Long_Delayed'] = BusBreakdown_df_clean['How_Long_Delayed'].str.extract('(\d+)')
BusBreakdown_df_clean['How_Long_Delayed'].head()

0    NaN
1     20
2     15
3     15
4    NaN
Name: How_Long_Delayed, dtype: object

We impute the null values in 'how_long_delayed' column with mean of the column. This will take a couple of steps...

In [ ]:
# We first convert our how_long_delayed to float type to allow for imputation
# ---
# 

BusBreakdown_df_clean['How_Long_Delayed']=BusBreakdown_df_clean['How_Long_Delayed'].apply(np.float64)


In [ ]:
# Then later perform our mean imputation 
# ---
# 
BusBreakdown_df_clean['How_Long_Delayed']=BusBreakdown_df_clean['How_Long_Delayed'].fillna(BusBreakdown_df_clean['How_Long_Delayed'].mean())


In [ ]:
# Then convert back our how_long_delayed column to integer datatype
# ---
# 
BusBreakdown_df_clean['How_Long_Delayed']=BusBreakdown_df_clean['How_Long_Delayed'].apply(np.int64)

In [ ]:
# Then check for nulls in the column
# ---
# 
BusBreakdown_df_clean['How_Long_Delayed'].isnull().sum()

0

In [ ]:
# Rename Boro column to Borough
# ---
# 
BusBreakdown_df_clean= BusBreakdown_df_clean.rename(columns= {'Boro' :'Borough'})
BusBreakdown_df_clean.columns

Index(['School_Year', 'Busbreakdown_ID', 'Run_Type', 'Bus_No', 'Route_Number',
       'Reason', 'Schools_Serviced', 'Occurred_On', 'Created_On', 'Borough',
       'Bus_Company_Name', 'How_Long_Delayed', 'Number_Of_Students_On_The_Bus',
       'Has_Contractor_Notified_Schools', 'Has_Contractor_Notified_Parents',
       'Have_You_Alerted_OPT', 'Informed_On', 'Incident_Number',
       'Last_Updated_On', 'Breakdown_or_Running_Late', 'School_Age_or_PreK'],
      dtype='object')

In [ ]:
# Lastly we convert all values in our colums to lower case
# for ease of reading
# ---
# 
BusBreakdown_df_clean.columns=BusBreakdown_df_clean.columns.str.lower()
BusBreakdown_df_clean.columns

Index(['school_year', 'busbreakdown_id', 'run_type', 'bus_no', 'route_number',
       'reason', 'schools_serviced', 'occurred_on', 'created_on', 'borough',
       'bus_company_name', 'how_long_delayed', 'number_of_students_on_the_bus',
       'has_contractor_notified_schools', 'has_contractor_notified_parents',
       'have_you_alerted_opt', 'informed_on', 'incident_number',
       'last_updated_on', 'breakdown_or_running_late', 'school_age_or_prek'],
      dtype='object')

In [ ]:
# Check the first 5 record the cleaned dataset
# ---
#
BusBreakdown_df_clean.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]


## 5. Solution Implementation

Here we investigate the questions that would help craft our recommendations.

### 5.a) Questions

In [ ]:
# 1. Which bus companies that had the highest breakdowns?
# ---
#
breakdowns = BusBreakdown_df_clean.groupby(['bus_company_name']).count()

breakdowns.sort_values(by='busbreakdown_id', ascending=0)


school_year  ...  school_age_or_prek
bus_company_name                                 ...                    
G.V.C., LTD.                              19394  ...               19394
LEESEL TRANSPORTATION CORP (B2192)        17200  ...               17200
RELIANT TRANS, INC. (B232                 13741  ...               13741
PIONEER TRANSPORTATION CO                 12017  ...               12017
BORO TRANSIT, INC.                        11953  ...               11953
...                                         ...  ...                 ...
SMART PICK INC                                1  ...                   1
R & C TRANSIT, INC. (B2321)                   1  ...                   1
L&M Bus Corp.                                 1  ...                   1
FORTUNA BUS COMPANY                           1  ...                   1
phillip bus service                           1  ...                   1

[117 rows x 20 columns]

In [ ]:
# 2. What were the top 3 reasons for bus delays?
# ---
#
bus_delay = BusBreakdown_df_clean.groupby(['reason']).count()[['how_long_delayed']].reset_index()
bus_delay

# Sort to get the most frequent reason

bus_delay.sort_values(by='how_long_delayed', ascending=0)




reason  how_long_delayed
3                Heavy Traffic            173221
6                        Other             37579
5           Mechanical Problem             28162
9                  Won`t Start             12283
2                    Flat Tire              8411
8           Weather Conditions              6932
4  Late return from Field Trip              5704
7                  Problem Run              4068
0                     Accident              2526
1            Delayed by School              2222

In [ ]:
# 3. How many students were in the buses when they broke down? 
# ---
BusBreakdown_df_clean.groupby(['number_of_students_on_the_bus']).count()[['busbreakdown_id']].reset_index()


number_of_students_on_the_bus  busbreakdown_id
0                                0           170376
1                                1            15704
2                                2            16737
3                                3            14837
4                                4            12248
..                             ...              ...
219                           6209                1
220                           6219                1
221                           8547                1
222                           9007                1
223                           9052                1

[224 rows x 2 columns]

In [ ]:
# 4. Which were most frequent reasons for bus breakdowns?
# ---

breakdown_reasons = BusBreakdown_df_clean[BusBreakdown_df_clean.breakdown_or_running_late == 'Breakdown'].groupby(['reason']).count()


# Sort to get most frequent reasons
breakdown_reasons.sort_values(by='busbreakdown_id', ascending=0)




school_year  ...  school_age_or_prek
reason                                    ...                    
Mechanical Problem                 14985  ...               14985
Won`t Start                         7731  ...                7731
Flat Tire                           4038  ...                4038
Other                               3530  ...                3530
Heavy Traffic                        419  ...                 419
Accident                             202  ...                 202
Weather Conditions                    82  ...                  82
Late return from Field Trip           35  ...                  35
Problem Run                           28  ...                  28
Delayed by School                      7  ...                   7

[10 rows x 20 columns]

In [ ]:
# 5. What were the most frequent reasons for the bus running late? 
# ---
Runninglate = BusBreakdown_df_clean[BusBreakdown_df_clean.breakdown_or_running_late == 'Running Late'].groupby(['reason']).count()

# Sort to get most frequent reasons
Runninglate.sort_values(by='busbreakdown_id', ascending=0)


school_year  ...  school_age_or_prek
reason                                    ...                    
Heavy Traffic                     172802  ...              172802
Other                              34049  ...               34049
Mechanical Problem                 13177  ...               13177
Weather Conditions                  6850  ...                6850
Late return from Field Trip         5669  ...                5669
Won`t Start                         4552  ...                4552
Flat Tire                           4373  ...                4373
Problem Run                         4040  ...                4040
Accident                            2324  ...                2324
Delayed by School                   2215  ...                2215

[10 rows x 20 columns]

In [ ]:
# 6. What was the average delay time for each reason type?
# ---

avg_delay = BusBreakdown_df_clean.groupby('reason').mean().reset_index()

# Get the records with reasons and how long on average a delay took then sort 

reason_delay_avg=avg_delay[['reason', 'how_long_delayed']]

reason_delay_avg.sort_values(by='how_long_delayed', ascending=0)


reason  how_long_delayed
4  Late return from Field Trip         62.235975
0                     Accident         34.965162
7                  Problem Run         31.136676
5           Mechanical Problem         30.915205
9                  Won`t Start         29.644061
2                    Flat Tire         29.392343
8           Weather Conditions         29.227351
6                        Other         28.421006
3                Heavy Traffic         26.225550
1            Delayed by School         18.907741

### 5.b) Recommendations

From the above analysis, below are our recommendations:

1.   Late return from Field Trip is the major cause of delays
2.   Heavy traffic is the major cause of buses running late
3.  Mechanical problems causes the most bus breakdowns

4.G.V.C., LTD. Bus company had the most recorded breakdowns




## 6. Challenging your Solution

During this step, we review our solution and implement  approaches that could potentially provide a better outcome. In our case, we could propose the following question that wasn't answered in our solution because it couldn't have greatly contributed to our recommendation.

In [ ]:
# Which boroughs experienced the most breakdowns?
# ---  
#
breakdowns = BusBreakdown_df_clean.groupby(['borough']).count()

# Sort to get boroughs with highest breakdowns
# ---
breakdowns.sort_values(by='busbreakdown_id', ascending=0)
#

school_year  ...  school_age_or_prek
borough                       ...                    
Bronx                  70952  ...               70952
Manhattan              68177  ...               68177
Brooklyn               62239  ...               62239
Queens                 38060  ...               38060
Staten Island          13907  ...               13907
Westchester             7093  ...                7093
Nassau County           4124  ...                4124
New Jersey              1583  ...                1583
Rockland County          875  ...                 875
All Boroughs             429  ...                 429
Connecticut              210  ...                 210

[11 rows x 20 columns]

Our observations:

1.  Bronx Borough has the most breakdowns  






## 7. Follow up questions

### a). Did we have the right data?

Yes, the data was right

### b). Do we need other data to answer our question?

Yes

### c). Did we have the right question?

Yes